kivy/eventmanager/__init__.py

```diff
@@ -0,0 +1,168 @@
+- Combine several motion events (touch, hover etc.) into one new event and
+  dispatch it to the widgets.
+- Dispatch one-time generic events, like app pause/resume.
+- Write an event simulator, like a touch simulator which draws a circle on
+  window's canvas for every simulated touch.
+
+Defining and registering an event manager
+-----------------------------------------
+
+1. Inherit :class:`EventManagerBase` and set which events this manager
+    should receive by declaring event types in
+    `:attr:`EventManagerBase.type_ids` attribute.
+2. Implement :meth:`EventManagerBase.dispatch` which will be called by window
+    to pass event type (one of "begin", "update", "end") and an event.
+3. Implement :meth:`EventManagerBase.start` and :meth:`EventManagerBase.stop`
+    to allocate and release additional resources if needed.
+4. Register a manager instance to window using method
+    :meth:`~kivy.core.window.WindowBase.register_event_manager`. This can be
+    done by overriding methods :meth:`~kivy.app.App.build` or
+    :meth:`~kivy.app.App.on_start`.
+
+All registered managers are kept in
+:attr:`~kivy.core.window.WindowBase.event_managers` list. To unregister a
+manager call :meth:`~kivy.core.window.WindowBase.unregister_event_manager`
+which itself can be called in :meth:`~kivy.app.App.on_stop` method.
+
+Dispatching events to the widgets
+---------------------------------
+
+Once registered, window will start the manager and forward all events of types
+declared in :attr:`EventManagerBase.type_ids` to the manager's
+:meth:`EventManagerBase.dispatch` method. It's up to manager to decide how to
+dispatch them, either going through :attr:`EventManagerBase.window.children`
+and dispatch `on_motion` event to them or by using some different logic. It's
+also up to manager to dispatch grabbed events if widgets grab/ungrab to events
+(see :meth:`~kivy.input.motionevent.MotionEvent.grab` and
+:meth:`~kivy.input.motionevent.MotionEvent.ungrab` methods).
+
+Manager can suggest how event should be dispatched further down the widget tree
+by using different flags. Flags are assigned in
+:attr:`~kivy.input.motionevent.MotionEvent.flags` attribute and inspected in
+:meth:`~kivy.uix.widget.Widget.on_motion` method by using bitwise operations.
+
+Currently there are three dispatch behaviors recognized by the widget:
+
+1. Default dispatch (no flag required) - event will go through widget's
+    `children` list, starting with fist widget in the `children` list until
+    child widget accept the event or last widget registered for that event is
+    reached.
+2. Filtered dispatch (requires :attr:`FILTERED_DISPATCH`) - event will go only
+    through registered child widgets.
+3. No children dispatch (requires :attr:`DONT_DISPATCH`) - event will not be
+    dispatched to child widgets.
+
+Note that window does not have `motion_filter` property and therefore does not
+have a list of filtered widgets from its `children` list.
+'''
+
+DONT_DISPATCH = 1
```

Comment:
Should we prefix those flag with `FLAG_`?

Goal with flags is to control dispatch behavior which can be different for the same event between manager and through widget tree between widgets. And then it's still open for user to add some of its own information (using bitwise operations) which can be tested by the widgets during dispatch. 

Maybe this is a better approach:
```
FLAG_DONT_DISPATCH = 2 ** 0
FLAG_DEFAULT_DISPATCH = 2 ** 1
FLAG_FILTERED_DISPATCH = 2 ** 2

FLAG_NEXT_EXP = 3 # Next exponent for user to use it its own app
```

This would imply that we would have to `from kivy.eventmanager import FLAG_DEFAULT_DISPATCH` in `motionevent.py` to set default value for flags https://github.com/kivy/kivy/pull/7658/files#diff-4a96e3b9598a90fae064bd5bee256a7ddc1b384f79b2c668a0382441979dbb0bR214.
Also to be explicit we should test for default flag on this line https://github.com/kivy/kivy/pull/7658/files#diff-7bb2215580b9e85b55ed03ae10df3a39914c7db316b996445f2ec40f4af9df7dR558 instead of using `else`.

Reply:
That's fine like this, but nothing against a `FLAG_` prefix if you find it clearer.

Reply:
> That's fine like this, but nothing against a FLAG_ prefix if you find it clearer.

In my opinion it would be clearer as when I read FLAG_ I assume bitwise operations.

> it doesn't make sense to set the value to DONT_DISPATCH + DEFAULT_DISPATCH

I think you meant `DONT_DISPATCH | DEFAULT_DISPATCH`, but anyway I agree, it doesn't make sense with currently available flags, but this is meant to be scalable so user can pass more information to widget and for framework can add more flags if needed in the next releases.

Reply:
Do we need `..versionadded:: 2.1.0` for flags?

Reply:
Yes, if assigned DONT_DISPATCH will be tested first, but that's only in the `Widget` class and subclasses can test for other user defined flags.

But now I think we should remove bitwise operations and just use assign "=" and test for equality "==". Because these flags cannot be combined, to have a combined meaning, then it would be better to just rename `MotionEvent.flags` to
`MotionEvent.dispatch_mode` and assign one of those 3 values. This will be more clear to users and it doesn't require from them to know bitwise operations. What do you think?

Reply:
That seems indeed clearer to me, and since the API is experimental, it can be changed later if we discover uses cases that would really benefit from the change, making the added complexity worthwhile, so yeah, better start with that :).

Reply:
Done in https://github.com/kivy/kivy/pull/7658/commits/bb87615c31742053c4623b12169706cc0fe16c74.

Is ":attr:\`MODE_FILTERED_DISPATCH\`" correct way to reference a constant in a module's docstring?

Reply:
Per sphinx documentation `const` should be used to reference a constant https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#role-py-const , so correct reference would be ":const:\`MODE_FILTERED_DISPATCH\`"

Reply:
It all done. You can squash and merge.